src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -172,19 +173,22 @@ public function set($id, $service)
             unset($this->aliases[$id]);
         }
 
-        $this->services[$id] = $service;
-
-        if (null === $service) {
+        if (null === $service) { // safeguard for 4.0
```

Comment:
As long as we stick with `protected $services` (opposed to private) i tend to keep this safeguard in 4.0, for historic reasons i guess. It would be better to remove this (adhere to the doc, like always) and remove all array_key_exists thingy's in 4.0